### PR TITLE
Optimization by replacing `%` in cycles

### PR DESCRIPTION
**Pull request by Nazarwadim**
I replaced the complex `%` with simple conditions, or by moving the body of the loop outside the loop. This does not seem to be optimized by the compiler, https://godbolt.org/z/6bW9v38dG.

**Benchmarks**
Used this https://quick-bench.com.
Clang 17
image

GCC 13.2
image

### CODE REVIEW

File: core/math/geometry_3d.cpp [hunk 1]
```diff
@@ -629,7 +629,10 @@ Geometry3D::MeshData Geometry3D::build_convex_mesh(const Vector<Plane> &p_planes
 			}
 
 			for (uint32_t k = 0; k < vertices.size(); k++) {
-				int k_n = (k + 1) % vertices.size();
+				uint32_t k_n = k + 1;
```
Comment (Capewearer) [inline core/math/geometry_3d.cpp:632]: Why uint32_t instead of int?

Comment (Nazarwadim) [reply core/math/geometry_3d.cpp:632]: To avoid a warning about comparing signed, unsigned int. And also `k` is a `uint32_t`.

**Comment by DarioSamo** [comment]
I can't speak much for the rest of the files but I'm not a fan of the changes to Rendering Device related classes. Most of these introduce a branch that didn't exist before, which makes the optimization claim require some profiling to back it up that it improves upon it. Branches break CPU pipelines far more than instructions like integer divisions.

Some of these are also replacing one liners for 3-4 lines of code in areas that are not hot paths at all. A few of the ones changed are either executed only once per frame or even during teardown.

I definitely recommend going for a bit more of a profiling-backed approach here as code maintenance is a valid concern. If some of these are identified to actually cause any significant performance problems (or you can show them with artificial tests), then optimizing with this approach will yield better results.

**Comment by Nazarwadim** [comment]
Well, first of all, I think that compilers are quite smart and can arrange branches correctly.
Also, smart clang does not even create branches, but uses this instructions.
```
cmp     rcx, 
mov     r9d, 
cmove   r9d, esi
```

And why I did it for non-hot spots is because it's much better to understand what's going on with the variable with a condition. At least for me.
``` c++
frame = (frame + 1) % frames.size();
```
```c++
frame++;
if (frame == frames.size()) {
	frame = 0;
}
```

**Comment by DarioSamo** [comment]
> Well, first of all, I think that compilers are quite smart and can arrange branches correctly.

My point is we don't need to rely on theory. You can just profile it and see if that holds true or not. Godot goes through many compilers so you can't rely on the behavior of just one of them. I'd rather go with an approach that I know that will work rather than trusting the compiler to get it right.

> And why I did it for non-hot spots is because it's much better to understand what's going on with the variable with a condition. At least for me.

That's a lot of code other people wrote intentionally that way that's gone through review and they found it to be clear that way. Otherwise they wouldn't have written it like that in the first place. Modulo is a very common operator and at least for positive numbers the operation is well defined. Removing the operator from _everywhere_, even in cases where it's used correctly, seems misguided to me.

**Review by ellenhp — COMMENTED** [review]
I don't think this should be merged. Some of these changes seem pretty error-prone, and it introduces some code-duplication that I don't like. Also, indexing into arrays in an unsafe language like C++ is really important to get right, and a lot of the code being changed is fairly well tested at this point. I don't think we should change it without good reason. As an end-user I'd rather have a working product that's a bit slower than an unstable (or worse: remotely exploitable) product that's imperceptibly faster, and I'm not even sure this will end up being faster. It's hard to say without profiling data.

**Comment by ellenhp** [comment]
I also want to say that I appreciate the contribution even though I don't think it should be merged. I saw you're a CS student on your GH profile and it makes me really happy to see students contributing stuff to big projects like Godot. :)

**Comment by DarioSamo** [comment]
> I also want to say that I appreciate the contribution even though I don't think it should be merged. I saw you're a CS student on your GH profile and it makes me really happy to see students contributing stuff to big projects like Godot. :)

Same goes for me!

**Comment by Nazarwadim** [comment]
Thank you :). I think we have to make a compromise. These changes definitely improve performance and are not optimized compilers and later I will give some benchmarks of these and other similar ones https://godbolt.org/z/6bW9v38dG. 

I will most likely revert everything back in many places and leave it only where we are in several loops and well covered by tests. 

As for the repetitive code, it has the greatest impact on performance, as you can see from the link I posted above, the compiler will be able to use `SIMD` registers, or if it already uses them, the number of `asm` commands will be much smaller. In some cases, when it is impossible to use `SIMD`, it will simply remove a heavy `mod` instructions from a loop that can be called millions of times.

**Comment by ellenhp** [comment]
While very useful for getting insight into the inner workings of compiled code, godbolt is not the same as, nor is it a substitute for a benchmark or a profiling result. It's impossible to know which is of these 3 functions is faster without actually running them, and your result will only be valid for that microarchitecture. What's faster on e.g. Broadwell is not always faster on all platforms that godot targets, so it's important to benchmark on a variety of platforms to justify a change like this. Modern CISC superscalar processors do not sequentially execute instructions, and on x86 especially there are some instructions that take _much_ longer than others, so you can't ever tell what's going to be faster just by looking at godbolt.

Even if you personally designed the processor that this code was running on, I would not believe that you could just guess which was faster by looking at the disassembly, because weighing a branch vs a mod operation is the epitome of a tradeoff that depends on the input data to the algorithm. Unless you can predict with relative certainty the state of the branch predictor at each branch, you don't know how long the pipeline will be stalled for.

Moreover, saving a few cycles here or there is less impactful than other things maintainers could spend time reviewing (especially bugfixes, bugfixes are super important), so I think you'll have a very hard time getting something like this merged unless it speeds up a major component of the engine in a way that's noticeable for users.

**Comment by ellenhp** [comment]
As an example of what a real-world case might look like with different input data, I ran quickbench after switching `val` to be `10` instead of the length of `ret`.

E-i2KDNEk1JBR9q4bmbXgJXzjKo

```
static void func_bad(benchmark::State& state) {
  // Code inside this loop is measured repeatedly
  for (auto _ : state) {
    int ret[10000];
    int val = 10;
    for(int i = 0; i < val; i++) {
        ret[i] = (i + 1) % val;
    }
    int sum = 0;
    for (int i = 0; i < val - 1; i++) {
        sum += ret[i];
    }
    benchmark::DoNotOptimize(sum);
  }
}
BENCHMARK(func_bad);

static void func_good(benchmark::State& state) {
  // Code inside this loop is measured repeatedly
  for (auto _ : state) {
    int ret[10000];
    int val = 10;
    for(int i = 0; i < val; i++) {
        int i_next = i + 1;
        if(i_next == val){
            i_next = 0;
        }
        ret[i] = i_next;
    }
    int sum = 0;
    for (int i = 0; i < val - 1; i++) {
        sum += ret[i];
    }
    benchmark::DoNotOptimize(sum);
  }
}
BENCHMARK(func_good);

static void func_very_good(benchmark::State& state) {
  // Code inside this loop is measured repeatedly
  for (auto _ : state) {
    int ret[10000];
    int val = 10;
    for(int i = 0; i < val - 1; i++) {
        ret[i] = i + 1;
    }
    ret[val - 1] = 0;
    int sum = 0;
    for (int i = 0; i < val - 1; i++) {
        sum += ret[i];
    }
    benchmark::DoNotOptimize(sum);
  }
}
BENCHMARK(func_very_good);
```

**Comment by Nazarwadim** [comment]
If you look at the benchmark I made (in the PR description), you can see a significant performance gain for code like this
```c++
			for (int i = 0; i < max_minus_one; i++) {
				iw[i * 2 + 0] = i;
				iw[i * 2 + 1] = i + 1;
			}
			iw[max_minus_one * 2 + 0] = max_minus_one;
			iw[max_minus_one * 2 + 1] = 0;
```
The compiler will not create branches with jumps, there will be much more difficult to understand `SIMD`. 
From the benchmark results `clang` doesn't see the difference between the first and second function, while `gcc` doesn't see the difference between the second and third. But `clang` makes the third option twice as fast as `gcc` does.

**Pull request closed without merging**